examples/nlp/text_generation_gpt.py

```diff
@@ -0,0 +1,418 @@
+    outputs = tokenizer(inputs)
+    features = outputs[:, :-1]
+    labels = outputs[:, 1:]
+    return features, labels
+
+
+# Tokenize and split into train and label sequences.
+ds = train_ds.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
+    tf.data.AUTOTUNE
+)
+ds_valid = valid_ds.map(preprocess, num_parallel_calls=tf.data.AUTOTUNE).prefetch(
+    tf.data.AUTOTUNE
+)
+
+"""
+## Model
+
+We create our scaled down GPT model with the following layers:
+
+- One `keras_nlp.layers.TokenAndPositionEmbedding` layer, which combines the embedding
+for the token and its position. 
+- Multiple `keras_nlp.layers.TransformerDecoder` layers, with the default causal masking.
+The layer has no cross-attention when run with decoder sequence only.
+- One output dense linear layer
+"""
+
+
+def create_model():
+    inputs = keras.layers.Input(shape=(SEQ_LEN,), dtype=tf.int32)
+    # Embedding.
+    embedding_layer = keras_nlp.layers.TokenAndPositionEmbedding(
+        vocabulary_size=VOCAB_SIZE,
+        sequence_length=SEQ_LEN,
+        embedding_dim=EMBED_DIM,
+        mask_zero=True,
+    )
+    x = embedding_layer(inputs)
+    # Transformer decoders.
+    for _ in range(NUM_LAYERS):
+        transformer_block = keras_nlp.layers.TransformerDecoder(
+            num_heads=NUM_HEADS,
+            intermediate_dim=FEED_FORWARD_DIM,
+        )
+        x = transformer_block(x)  # Giving one argument only skips cross-attention.
+    # Output.
+    outputs = keras.layers.Dense(VOCAB_SIZE)(x)
+    model = keras.Model(inputs=inputs, outputs=outputs)
+    loss_fn = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True)
+    perplexity = keras_nlp.metrics.Perplexity(from_logits=True, mask_token_id=0)
+    model.compile(optimizer="adam", loss=loss_fn, metrics=[perplexity])
+    return model
+
+
+model = create_model()
+
+"""
+Let's take a look at our model summary here! We can see that a large majority of the
+parameters are in the token and position embedding and the output dense layer! This means
+that the vocabulary size (`VOCAB_SIZE`) has a large affect on the size of the model,
+while the number of transformer decoder layers (`NUM_LAYERS`) doesn't affect it much.
+"""
+
+model.summary()
+
+"""
+## Training
+
+Now that we have our model, let's train it. We use a subset of the training data to save
+on training time. It would also be beneficial to use a GPU to speed up the training
+process!
+"""
+
+model.fit(ds.take(NUM_TRAINING_BATCHES), validation_data=ds_valid, verbose=2, epochs=EPOCHS)
+
+"""
+## Inference
+
+With our trained model, we can test it out to gauge it's performance. Since
+this is a dataset of mostly fictional books, there is bound to be a hero, so let's use
+"The hero" as our starting string! We run it through the tokenizer to get the input for
+the model. 
+"""
+
+MAX_PREDICT_LEN = 80
+start_prompt = "The hero"
+# Unpadded token sequence.
+start_tokens = [tokenizer.token_to_id(_) for _ in start_prompt.lower().split()]
```

Comment:
this feels awkward, we should actually use a tokenizer to tokenize the text. maybe let's just instantiate the tokenizer without a sequence length, and either use ragged.to_dense() or a packer layer to densify (especially if we decide to add start tokens)

Reply:
edited to use packer layer for start tokens